src/java/org/apache/cassandra/cql3/constraints/AbstractFunctionSatisfiabilityChecker.java

```diff
@@ -0,0 +1,215 @@
+{
+    /**
+     * Performs check if constraints are satisfiable or not.
+     *
+     * @param functionName   name of function
+     * @param constraints    list of constraints to set
+     * @param columnMetadata metadata of a column.
+     */
+    public void check(String functionName, List<ColumnConstraint<?>> constraints, ColumnMetadata columnMetadata)
+    {
+        Pair<List<CONSTRAINT_TYPE>, List<CONSTRAINT_TYPE>> filteredConstraints = categorizeConstraints(constraints, functionName);
+
+        if (filteredConstraints.left.isEmpty())
+            return;
+
+        checkNumberOfConstraints(columnMetadata, filteredConstraints);
+        checkSupportedOperators(filteredConstraints.left, functionName);
+        ensureSatisfiability(columnMetadata, functionName, filteredConstraints.left);
+    }
+
+    /**
+     * Categorizes given constraints into two lists. The first list, the left one in Pair, contains all
+     * constraints of implementation-specific {@link org.apache.cassandra.cql3.constraints.ColumnConstraint.ConstraintType}.
+     * The second list, the right one in Pair, contains all constraints of such constraint type which do have "not equal" operator.
+     *
+     * @param constraints  constraints to categorize
+     * @param functionName name of function
+     * @return pair of categorized constraints
+     */
+    abstract Pair<List<CONSTRAINT_TYPE>, List<CONSTRAINT_TYPE>> categorizeConstraints(List<ColumnConstraint<?>> constraints, String functionName);
+
+    private void checkSupportedOperators(List<CONSTRAINT_TYPE> allConstraints, String functionName)
+    {
+        for (CONSTRAINT_TYPE constraint : allConstraints)
+        {
+            if (!constraint.getSupportedOperators().contains(constraint.relationType()))
+                throw new InvalidConstraintDefinitionException(format("%s constraint of relation '%s' is not supported.", functionName, constraint.relationType()));
+        }
+    }
+
+    /**
+     * Checks if there are no duplicate constraints having same operator.
+     *
+     * @param columnMetadata      medata of a column
+     * @param filteredConstraints pair of all constraints and all constraints having not-equal operator
+     */
+    private void checkNumberOfConstraints(ColumnMetadata columnMetadata, Pair<List<CONSTRAINT_TYPE>, List<CONSTRAINT_TYPE>> filteredConstraints)
+    {
+        List<? extends AbstractFunctionConstraint<CONSTRAINT_TYPE>> allConstraints = filteredConstraints.left;
+        List<? extends AbstractFunctionConstraint<CONSTRAINT_TYPE>> notEqualConstraints = filteredConstraints.right;
+
+        if ((allConstraints.size() - notEqualConstraints.size() > 2))
+        {
+            throw new InvalidConstraintDefinitionException(format("There can not be more than 2 constraints (not including non-equal relations) on a column '%s' but you have specified %s",
```

Comment:
This has me worried after the counter example we discussed offline:
```
alter ks.tb alter val check json(val) and length(val) < 1000 and length(val)>100
```

Should be a valid constraint.

Reply:
@bbotella that will be all possible. This check is for _one particular constraint function_.

So it will be treated for e.g. scalar function, then it will be treated for lenght specifically ... for any function, separately.

json function will have it own "satisfiability checker" and for that particular function, it is unary function, and it will be always satisfied. We do not need to look into any other function to decide if it is satisfiable or not.